File: src/main/java/net/volcano/jdacommands/model/command/annotations/CommandController.java
package net.volcano.jdacommands.model.command.annotations;

import net.volcano.jdacommands.model.command.Command;
import org.springframework.core.annotation.AliasFor;
import org.springframework.stereotype.Component;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Component
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
public @interface CommandController {
	
	@AliasFor("path")
	String[] value() default {""};
	
	/**
	 * @return base path for all commands under this controller
	 */
	@AliasFor("value")
	String[] path() default {""};
	
	/**
	 * @return base permissions required for any command under this controller
	 */
	String permissions() default "";
	
	/**
	 * @return Base source to apply to all commands under this controller unless they specify otherwise
	 */
	net.volcano.jdacommands.model.command.Command.Source source() default Command.Source.DEFAULT;
	
	/**
	 * The default category for this controller commands
	 */
	String category() default "general";
	
}

File: src/main/java/net/volcano/jdacommands/model/command/annotations/CommandController.kt
package net.volcano.jdacommands.model.command.annotations

import net.volcano.jdacommands.model.command.Source
import org.springframework.stereotype.Component

@Component
@Target(AnnotationTarget.ANNOTATION_CLASS, AnnotationTarget.CLASS)
@Retention(AnnotationRetention.RUNTIME)
annotation class CommandController(

	vararg val value: String = [""],

	/**
	 * @return base path for all commands under this controller
	 */
	val path: Array<String> = [""],

	/**
	 * @return base permissions required for any command under this controller
	 */
	val permissions: String = "",

	/**
	 * @return Base source to apply to all commands under this controller unless they specify otherwise
	 */
	val source: Source = Source.DEFAULT,

	/**
	 * The default category for this controller commands
	 */
	val category: String = "general"

)